merkletree/pad.go

```diff
@@ -76,48 +92,77 @@ func (pad *PAD) Update(policies Policies) {
 		pad.loadedEpochs = append(pad.loadedEpochs[:0], pad.loadedEpochs[n:]...)
 	}
 
-	pad.updateInternal(policies, pad.currentSTR.epoch+1)
+	pad.updateInternal(policies, pad.latestSTR.epoch+1)
 }
 
 func (pad *PAD) Set(key string, value []byte) error {
-	return pad.tree.Set(key, value)
+	index, _ := pad.computePrivateIndex(key, pad.policies.vrfPrivate())
+	return pad.tree.Set(index, key, value)
 }
 
-func (pad *PAD) Lookup(key string) *AuthenticationPath {
-	str := pad.currentSTR
-	return str.tree.Get(key)
+func (pad *PAD) Lookup(key string) (*AuthenticationPath, error) {
+	return pad.LookupInEpoch(key, pad.latestSTR.epoch)
 }
 
 func (pad *PAD) LookupInEpoch(key string, epoch uint64) (*AuthenticationPath, error) {
 	str := pad.GetSTR(epoch)
 	if str == nil {
 		return nil, ErrorSTRNotFound
 	}
-	ap := str.tree.Get(key)
+	lookupIndex, proof := pad.computePrivateIndex(key, str.policies.vrfPrivate())
+	ap := str.tree.Get(lookupIndex)
+	ap.vrfProof = proof
 	return ap, nil
 }
 
 func (pad *PAD) GetSTR(epoch uint64) *SignedTreeRoot {
-	if epoch >= pad.currentSTR.epoch {
-		return pad.currentSTR
+	if epoch >= pad.latestSTR.epoch {
+		return pad.latestSTR
 	}
 	return pad.snapshots[epoch]
 }
 
 func (pad *PAD) TB(key string, value []byte) (*TemporaryBinding, error) {
-	//FIXME: compute private index twice
-	//it would be refactored after merging VRF integration branch
-	index := computePrivateIndex(key)
-	tb := pad.currentSTR.sig
+	str := pad.latestSTR
+	if str == nil {
+		panic(ErrorSTRNotFound)
+	}
+	index, _ := pad.computePrivateIndex(key, pad.policies.vrfPrivate())
+	tb := str.sig
 	tb = append(tb, index...)
 	tb = append(tb, value...)
 	sig := crypto.Sign(pad.key, tb)
 
-	err := pad.Set(key, value)
+	err := pad.tree.Set(index, key, value)
 
 	return &TemporaryBinding{
 		index: index,
 		value: value,
 		sig:   sig,
 	}, err
 }
+
+func (pad *PAD) reshuffle() error {
+	var err error
+	newTree, err := NewMerkleTree()
+	if err != nil {
+		return err
+	}
+	pad.tree.visitLeafNodes(func(n *userLeafNode) {
+		newIndex, _ := pad.computePrivateIndex(n.key, pad.policies.vrfPrivate())
+		if e := newTree.Set(newIndex, n.key, n.value); e != nil {
+			// or should we panic here?
```

Comment:
I think we should panic here, since we may have many same errors returned by `Set()`. In that case, we probably want to have some ways to recovery it ASAP. Hope it makes sense to you.